xapian-applications/omega/postprocess

```diff
@@ -0,0 +1,247 @@
+
+            # Serialise "Hits" and "Clicks"
+            row[HITS] = ','.join(row[HITS])
+            clicklist = ','.join(clicklist)
+
+            row.append(clicklist)
+            writer.writerow(row)
+
+
+def generate_query_file(final_log, query_file):
+    """Generates query file formatted as per Xapian Letor documentation.
+
+    Input Args:
+        final_log (string): Path to final.log file.
+        query_file (string): Path to query.txt file.
+
+    Example (comma-delimited) entries in final.log:
+
+    QueryID,Query,Hits,Offset,Clicks
+    821f03288846297c2cf43c34766a38f7,book,"45,54",0,"45:0,54:2"
+    098f6bcd4621d373cade4e832627b4f6,test,"35,47",0,"35:1,47:0"
+
+    Example (comma-delimited) entries in query.txt:
+
+    821f03288846297c2cf43c34766a38f7,book
+    098f6bcd4621d373cade4e832627b4f6,test
+    """
+    with open(final_log, 'r') as s, open(query_file, 'w+') as w:
+        reader = csv.DictReader(s)
+        writer = csv.writer(w)
+
+        for row in reader:
+            writer.writerow([row['QueryID'], row['Query']])
+
+
+def test_functions():
+    with open('test_search.log', 'w+') as test_search:
+        test_search.write('821f03288846297c2cf43c34766a38f7,"book","45,54",0\n')
+        test_search.write('821f03288846297c2cf43c34766a38f7,"book","45,54",0\n')
+        test_search.write('d41d8cd98f00b204e9800998ecf8427e,"","",0\n')
+        test_search.write('098f6bcd4621d373cade4e832627b4f6,"test","35,47",0\n')
+
+    with open('test_clicks.log', 'w+') as test_clicks:
+        test_clicks.write('821f03288846297c2cf43c34766a38f7,54\n')
+        test_clicks.write('821f03288846297c2cf43c34766a38f7,54\n')
+        test_clicks.write('098f6bcd4621d373cade4e832627b4f6,35\n')
+
+    generate_combined_log('test_search.log', 'test_clicks.log', 'test_final.log')
+
+    # Test entries in final.log are correct.
+    with open('test_final.log', 'r') as final_f:
+        reader = csv.reader(final_f)
+
+        # Skip header row.
+        next(final_f)
+
+        for i, row in enumerate(reader):
+            if i == 0:
+                assert row == ['821f03288846297c2cf43c34766a38f7',
+                              'book', '45,54', '0', '45:0,54:2'], "Incorrect entry in final.log"
+            if i == 1:
+                assert row == ['098f6bcd4621d373cade4e832627b4f6',
+                              'test', '35,47', '0', '35:1,47:0'], "Incorrect entry in final.log"
+
+    generate_query_file('test_final.log', 'test_query.txt')
+
+    # Test entries in query.txt are correct.
+    with open('test_query.txt', 'r') as query_f:
+        reader = csv.reader(query_f)
+
+        for i, row in enumerate(query_f):
+            if i == 0:
+                assert row == '821f03288846297c2cf43c34766a38f7,book\n', "Incorrect entry in query.txt"
+            if i == 1:
+                assert row == '098f6bcd4621d373cade4e832627b4f6,test\n', "Incorrect entry in query.txt"
+
+    # Safely delete test log files.
```

Comment:
These need to happen as a `finally` in case the earlier code raises an exception (eg by failing an `assert`).